flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java

```diff
@@ -396,14 +397,16 @@ private TaskExecutorFailureStrategy(int failAfterCallNumber) {
 		@Override
 		void fail(int trackingIndex) throws Exception {
 			//noinspection OverlyBroadCatchBlock
-			try {
-				restartTaskManager();
-			} catch (InterruptedException e) {
-				// ignore the exception, task should have been failed while stopping TM
-				Thread.currentThread().interrupt();
-			} catch (Throwable t) {
-				failureTracker.unrelatedFailure(t);
-				throw t;
+			try (TemporaryClassLoaderContext unused = TemporaryClassLoaderContext.of(ClassLoader.getSystemClassLoader())) {
```

Comment:
why do we need a separate classloader here?

Reply:
The test is badly written, but I haven't found a better solution. We are restarting a task manager from within a UDF (with context classloader set to user class loader). Restarting means closing TM, which closes context class loader, such that starting the TM fails.